File: data_structure/array/240_search_a_2d_matrix_ii.cc
#include <iostream>
#include <vector>

// 关键: 从右上角或者左下角开始查找
// 具体实现并不难.
bool SearchMatrix(const std::vector<std::vector<int>>& matrix, int target) {
  const int m = matrix.size();
  const int n = matrix[0].size();
  int i = 0, j = n - 1;
  while (i < m && j >= 0) {
    if (matrix[i][j] < target) {
      ++i;
    } else if (matrix[i][j] > target) {
      --j;
    } else {
      return true;
    }
  }
  return false;
}

int main(int argc, char* argv[]) {
  std::vector<std::vector<int>> matrix = {{1, 4, 7, 11, 15},
                                          {2, 5, 8, 12, 19},
                                          {3, 6, 9, 16, 22},
                                          {10, 13, 14, 17, 24},
                                          {18, 21, 23, 26, 30}};
  std::cout << SearchMatrix(matrix, 2) << std::endl;
  return 0;
}

File: data_structure/array/240_search_a_2d_matrix_ii.py
from typing import List


# 2026/05/12: 自己实现
def search_matrix(matrix: List[List[int]], target: int) -> bool:
    if not matrix or not matrix[0]:
        return False

    i, j = 0, len(matrix[0]) - 1
    while i < len(matrix) and j >= 0:
        if matrix[i][j] == target:
            return True
        elif matrix[i][j] < target:
            i += 1
        elif matrix[i][j] > target:
            j -= 1

    return False


matrix1 = [
    [1, 4, 7, 11, 15],
    [2, 5, 8, 12, 19],
    [3, 6, 9, 16, 22],
    [10, 13, 14, 17, 24],
    [18, 21, 23, 26, 30],
]
target1 = 5

matrix2 = [
    [1, 4, 7, 11, 15],
    [2, 5, 8, 12, 19],
    [3, 6, 9, 16, 22],
    [10, 13, 14, 17, 24],
    [18, 21, 23, 26, 30],
]
target2 = 2

print(search_matrix(matrix1, target1))
print(search_matrix(matrix2, target2))
